### DISCUSSION TITLE
curl_multi_exec with CURLOPT_IPRESOLVE connection reuse

### DISCUSSION THREAD

**Discussion by itxp2008** [curl/curl#10604] [General]
Hi,

Can someone guide me how to properly use curl_multi_exec with CURLOPT_IPRESOLVE and keep alive connection reuse?

I can also provide my php butchered code:(.

Basically I want to do some tests/benchmarks for a sever behind cloudfront. I've selected some individual "fast" cloudfront IPs and now I basically want to open 10 concurrent connections, one per IP, keep the connection alive and reuse it for the test request.

Basically I initialized the multi_handle and the containing curl handles with the appropriate IPs, requested a timestamp on the server.
This should handle opening the connection, ssl handshakes, etc.

Wait 5 seconds

After when I submit the 2nd batch to the same handles it uses the same 1st primary ip for each one of them.

The goal is to test the benefits of using concurrent confections trough cloudfront without other things, connect time, etc.

The 1st batch is just a dummy one to open the connection, in the second batch I'm interested after the connection is opened. Also if there's other possibility of opening the connections and after 5 seconds to submit the request.

Also if I can get the timestamp of each request sent by cul_multi_exec.

Regards

**Comment by bagder** [comment]
if you let those 10 separate connections use different host names then there's no problem, is it?

**Reply by itxp2008** [reply]
Unfortunately I cannot, the webserver just listens to this specific virtual host.

But i don't know what I'm missing because sending the 1st batch 10 requests(one per ip)

curl_info for primary_ip after running curl_multi_exec reports those 10 different IPs per each curl_handle

But how to reuse those connections is the thing because I don't care about the parallelization in the 1st batch, this is used just to open the connections, should I open them in a loop or something one per ip(talking about the curl_handles) then wait N seconds and then init the multihandle and execute? Now I thought about this as a bypass.

But there's not much info for this, again multi_exec on the 1st batch works but on the second just uses same 1st ip.

Don't know what I'm missing, I can share my code, but just pseudocode is enough for me in which order I should do this.

**Reply by bagder** [reply]
> But how to reuse those connections is the thing

curl's connection reuse is name-based so if you do another transfer to a host name to which curl already has a live connection for, that connection will be used. Independently of which IP address it has or uses.

**Reply by itxp2008** [reply]
interesting, I thought those curl_handles act independently, I will do a extra test, without curl_multi altogether, neither for opening the connections and then to submit the requests in a curl_exec loop... just to test your sayings and learn, then I will split them somehow, multithreading, subprocesses, new processes, forked, whatever. But I've seen that curl_multi is faster for ex. 10 parallel connections then running curl_exec multithreaded/...

any suggestions on how to perform this test in parallel, cloudfront in question has a ton of IPs and a ton of routes behind the scenes, I've compiled a list of "fast" IPs that have the best route behind them to the server. I cannot rely on just one at a moment in time and the test is relying on logic which ever comes first is the winner. Imagine 10 routes between 2 cities, you know from the start that one is shorter then the other, but on the way, at the time of reference after sending 10 cars(depart ~same time) one arrives faster on another(on other route then the shortest).

This test can prove the consistency of my hypothesis.

Thank you for your support.

**Reply by itxp2008** [reply]
Done some tests,

1. test1

for_loop different ips curl_exec same host

CURLINFO_PRIMARY_IP checked and it works with multi IPs

wait 5 sec

again just curl_exec again in a for_loop

CURLINFO_PRIMARY_IP checked and it works with multi IPs

each curl_handle reports it's own used IP and the connection was kept_alive/reused

2. test 2

for_loop different ips curl_exec same host

CURLINFO_PRIMARY_IP checked and it works with multi IPs

wait 5 sec

initialized the multi curl handle with these for reuse

curl_multi_exec

CURLINFO_PRIMARY_IP reports same IP from the 1st connection and the timings suggest that the connection wasn't reused, but was take from the start with app_connect/other overhead for every connection

It gets frustrating, thought that curl_multi_exec will act in the same manner as curl_exec but in parallel.

**Reply by itxp2008** [reply]
I don't know at this moment if it's a matter of bad usage of curl_multi_exec or this discussion should transform into an issue.

Doing curl_exec 10 x curl handles, 1st batch to open the connection, wait and then 2nd batch on those connections, curl_exec respects the connections and the primary_ip allocated to each one of them.

But doing curl_multi_exec doesn't respect the different primary_ips used in the handles.

**Reply by bagder** [reply]
I don't understand that "respecting" IPs mean. I already explained how connection re-use is name-based. It matches the host names and will try to re-use a connection if there is a live connection to the host. It then checks the name and does not care about the IP. curl and libcurl are transfer oriented, meant to perform transfers as fast and convenient as possible, which at times makes them less suitable for custom performance measurements.

**Reply by itxp2008** [reply]
> I don't understand that "respecting" IPs mean

I meant corresponding IPs.

>  I already explained how connection re-use is name-based.

I don't want to contradict you, just want to understand.
I've done a test with curl_exec in a for loop and it works. Same hostname, connection reuse, for those parallel connections that we're opened prior to the 2nd batch, and it works.

`$ips = [
    1 => '13.35.68.47',
    2 => '13.35.46.211',
    3 => '13.227.49.232',
    4 => '99.84.61.17',
    5 => '13.35.68.47',
    6 => '13.35.46.211',
    7 => '13.227.49.232',
    8 => '99.84.61.17',
];

for($i=1; $i<=8; $i++){

    $fetchURL = 'https://api.example.com/server_time';
    $multiCurl[$i] = curl_init();
    curl_setopt($multiCurl[$i], CURLOPT_URL,$fetchURL);
    curl_setopt($multiCurl[$i], CURLOPT_HEADER,1);
    curl_setopt($multiCurl[$i], CURLOPT_RETURNTRANSFER,1);
    curl_setopt($multiCurl[$i], CURLOPT_RESOLVE , ['api.example.com:443:'.$ips[$i]] );

}

foreach($multiCurl as $k => $ch) {

    $response = curl_exec($ch);
    //get headers #1 here
    //get curl_info #1 here
    //get responses #1 here
}

sleep(5);

foreach($multiCurl as $k => $ch) {

    $response = curl_exec($ch);

    //get headers #2 here
    //get curl_info #2 here
    //get responses #2 here

}`

curl_infos1 =

 {
    "1": {
        "CURLINFO_PRIMARY_IP": "13.35.68.47",
        "CURLINFO_TOTAL_TIME": 0.05478,
        "CURLINFO_CONNECT_TIME": 0.000279,
        "CURLINFO_NUM_CONNECTS": 1,
        "CURLINFO_TOTAL_TIME_T": 54780,
        "CURLINFO_EFFECTIVE_URL": "https:\/\/api.example.com\/server_time",
        "CURLINFO_CONNECT_TIME_T": 279,
        "CURLINFO_APPCONNECT_TIME": 0.051559,
        "CURLINFO_NAMELOOKUP_TIME": 2.2e-5,
        "CURLINFO_PRETRANSFER_TIME": 0.051632,
        "CURLINFO_APPCONNECT_TIME_T": 51559,
        "CURLINFO_NAMELOOKUP_TIME_T": 22,
        "CURLINFO_PRETRANSFER_TIME_T": 51632,
        "CURLINFO_STARTTRANSFER_TIME": 0.054757,
        "CURLINFO_STARTTRANSFER_TIME_T": 54757
    },
    "2": {
        "CURLINFO_PRIMARY_IP": "13.35.46.211",
        "CURLINFO_TOTAL_TIME": 0.055168,
        "CURLINFO_CONNECT_TIME": 0.000246,
        "CURLINFO_NUM_CONNECTS": 1,
        "CURLINFO_TOTAL_TIME_T": 55168,
        "CURLINFO_EFFECTIVE_URL": "https:\/\/api.example.com\/server_time",
        "CURLINFO_CONNECT_TIME_T": 246,
        "CURLINFO_APPCONNECT_TIME": 0.052272,
        "CURLINFO_NAMELOOKUP_TIME": 1.5e-5,
        "CURLINFO_PRETRANSFER_TIME": 0.052327,
        "CURLINFO_APPCONNECT_TIME_T": 52272,
        "CURLINFO_NAMELOOKUP_TIME_T": 15,
        "CURLINFO_PRETRANSFER_TIME_T": 52327,
        "CURLINFO_STARTTRANSFER_TIME": 0.055149,
        "CURLINFO_STARTTRANSFER_TIME_T": 55149
    },
    "3": {
        "CURLINFO_PRIMARY_IP": "13.227.49.232",
        "CURLINFO_TOTAL_TIME": 0.056655,
        "CURLINFO_CONNECT_TIME": 0.000273,
        "CURLINFO_NUM_CONNECTS": 1,
        "CURLINFO_TOTAL_TIME_T": 56655,
        "CURLINFO_EFFECTIVE_URL": "https:\/\/api.example.com\/server_time",
        "CURLINFO_CONNECT_TIME_T": 273,
        "CURLINFO_APPCONNECT_TIME": 0.051482,
        "CURLINFO_NAMELOOKUP_TIME": 2.1e-5,
        "CURLINFO_PRETRANSFER_TIME": 0.051533,
        "CURLINFO_APPCONNECT_TIME_T": 51482,
        "CURLINFO_NAMELOOKUP_TIME_T": 21,
        "CURLINFO_PRETRANSFER_TIME_T": 51533,
        "CURLINFO_STARTTRANSFER_TIME": 0.056635,
        "CURLINFO_STARTTRANSFER_TIME_T": 56635
    },

......

and for the 2nd batch

curl_infos2 =
{
    "1": {
        "CURLINFO_PRIMARY_IP": "13.35.68.47",
        "CURLINFO_TOTAL_TIME": 0.003376,
        "CURLINFO_CONNECT_TIME": 7.1e-5,
        "CURLINFO_NUM_CONNECTS": 0,
        "CURLINFO_TOTAL_TIME_T": 3376,
        "CURLINFO_EFFECTIVE_URL": "https:\/\/api.example.com\/server_time",
        "CURLINFO_CONNECT_TIME_T": 71,
        "CURLINFO_APPCONNECT_TIME": 7.1e-5,
        "CURLINFO_NAMELOOKUP_TIME": 7.1e-5,
        "CURLINFO_PRETRANSFER_TIME": 0.000111,
        "CURLINFO_APPCONNECT_TIME_T": 71,
        "CURLINFO_NAMELOOKUP_TIME_T": 71,
        "CURLINFO_PRETRANSFER_TIME_T": 111,
        "CURLINFO_STARTTRANSFER_TIME": 0.003357,
        "CURLINFO_STARTTRANSFER_TIME_T": 3357
    },
    "2": {
        "CURLINFO_PRIMARY_IP": "13.35.46.211",
        "CURLINFO_TOTAL_TIME": 0.003367,
        "CURLINFO_CONNECT_TIME": 2.4e-5,
        "CURLINFO_NUM_CONNECTS": 0,
        "CURLINFO_TOTAL_TIME_T": 3367,
        "CURLINFO_EFFECTIVE_URL": "https:\/\/api.example.com\/server_time",
        "CURLINFO_CONNECT_TIME_T": 24,
        "CURLINFO_APPCONNECT_TIME": 2.4e-5,
        "CURLINFO_NAMELOOKUP_TIME": 2.4e-5,
        "CURLINFO_PRETRANSFER_TIME": 4.0e-5,
        "CURLINFO_APPCONNECT_TIME_T": 24,
        "CURLINFO_NAMELOOKUP_TIME_T": 24,
        "CURLINFO_PRETRANSFER_TIME_T": 40,
        "CURLINFO_STARTTRANSFER_TIME": 0.003357,
        "CURLINFO_STARTTRANSFER_TIME_T": 3357
    },
    "3": {
        "CURLINFO_PRIMARY_IP": "13.227.49.232",
        "CURLINFO_TOTAL_TIME": 0.002868,
        "CURLINFO_CONNECT_TIME": 2.0e-5,
        "CURLINFO_NUM_CONNECTS": 0,
        "CURLINFO_TOTAL_TIME_T": 2868,
        "CURLINFO_EFFECTIVE_URL": "https:\/\/api.example.com\/server_time",
        "CURLINFO_CONNECT_TIME_T": 20,
        "CURLINFO_APPCONNECT_TIME": 2.0e-5,
        "CURLINFO_NAMELOOKUP_TIME": 2.0e-5,
        "CURLINFO_PRETRANSFER_TIME": 3.4e-5,
        "CURLINFO_APPCONNECT_TIME_T": 20,
        "CURLINFO_NAMELOOKUP_TIME_T": 20,
        "CURLINFO_PRETRANSFER_TIME_T": 34,
        "CURLINFO_STARTTRANSFER_TIME": 0.00286,
        "CURLINFO_STARTTRANSFER_TIME_T": 2860
    },

Now you can see what I was talking about, it works, the connection is reused with it's own IP, but this with curl_multi doesn't work on the 2nd request, on the 1st it does.... and also you can see in the infos2 that the connection was reused.

Why curl_multi works on the 1st and on the 2nd it doesn't. Meanwhile curl_exec is more then capable to reuse multiple connections to the same host with different IPs.

With curl_multi the 1st batch looks the same, and on the 2nd it reuses a connection or something because of the timings but uses the same ip. The curl_opt are set in the curl_handle and not on the multi_handle, maybe is just that curl_multi_exec is not a multi curl_exec and the CURLOPT_RESOLVE used for each ch is for no reason, but again on the 1st is respected and on the second is not.

I'm thinking that is either a bad usage of curl_multi for conn reuse, didn't find anything online regarding this or maybe is an issue. I don't see either then one of these.

**Reply by itxp2008** [reply]
I really like the performance of curl_multi and their approach, now I have to go to promises/multithreading/subprocesses/forks

More complicated and each has it's own caveat

Getting process to sync and then find a way to signal all, fork has a delay, promises have higher exec time then curl_muti and s.o.
